lib/crypto/src/eddsa.rs

```diff
@@ -0,0 +1,375 @@
+    }
+
+    /// Verify a signature on a message with this signing key's public key.
+    #[must_use]
+    pub fn is_valid_signature(
+        &self,
+        message: &[u8],
+        signature: &Signature,
+    ) -> bool {
+        self.verifying_key.is_valid(message, signature)
+    }
+
+    /// Sign a message with this signing key's secret key.
+    ///
+    /// ```rust
+    ///  use openzeppelin_crypto::eddsa::{SecretKey, SigningKey, SECRET_KEY_LENGTH};
+    ///
+    ///  let secret_key: SecretKey = [1u8; SECRET_KEY_LENGTH];
+    ///  let signing_key = SigningKey::from_bytes(&secret_key);
+    ///  let message = b"Sign me!";
+    ///
+    ///  let signature = signing_key.sign(message);
+    ///  assert!(signing_key.is_valid_signature(message, &signature));
+    /// ```
+    #[must_use]
+    pub fn sign(&self, message: &[u8]) -> Signature {
+        let mut h = Sha512::new();
+
+        h.update(self.signing_key.hash_prefix);
+        h.update(message);
+
+        let r = WideScalar::from_bigint(U512::from_bytes_le(
+            h.finalize().as_slice(),
+        ));
+        let r = Scalar::from_fp(r);
+
+        let R = ProjectivePoint::generator() * r;
+
+        h = Sha512::new();
+        h.update(CompressedPointY::from(R.into_affine()));
+        h.update(CompressedPointY::from(
+            self.verifying_key.point.into_affine(),
+        ));
+        h.update(message);
+
+        let k = WideScalar::from_bigint(U512::from_bytes_le(
+            h.finalize().as_slice(),
+        ));
+        let k = Scalar::from_fp(k);
+        let s: Scalar = (k * self.signing_key.scalar) + r;
+
+        Signature { R, s }
+    }
+}
+
+/// In "Ed25519" format, the curve point `(x,y)` is
+/// determined by the y-coordinate and the sign of `x`.
+///
+/// The first 255 bits of a `CompressedEdwardsY` represent the `y`-coordinate.
+/// The high bit of the 32nd byte gives the sign of `x`.
+#[derive(Copy, Clone, Hash)]
+pub struct CompressedPointY([u8; 32]);
+
+impl AsRef<[u8]> for CompressedPointY {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl From<AffinePoint> for CompressedPointY {
+    fn from(point: AffinePoint) -> Self {
+        let mut s: [u8; 32] = point
+            .y
+            .into_bigint()
+            .into_bytes_le()
+            .try_into()
+            .expect("Y coordinate should be of 32 bit");
+
+        let is_odd = point.x.into_bigint().is_odd();
+        s[31] ^= u8::from(is_odd) << 7;
+
+        CompressedPointY(s)
+    }
+}
+
+/// Ed25519 signature representation.
+///
+/// Signature verification libraries are expected to reject invalid
+/// field elements at the time a signature is verified.
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct Signature {
+    /// `R` is an `EdwardsPoint`, formed by using an hash function with
+    /// `512-bits` output to produce the digest of:
+    ///
+    /// - the nonce half of the `ExpandedSecretKey`, and
+    /// - the message to be signed.
```

Comment:
```suggestion
    /// * the nonce half of the `ExpandedSecretKey`, and
    /// * the message to be signed.
```
nit: current convention